python/tchannel/tornado/data.py

```diff
@@ -0,0 +1,421 @@
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from __future__ import absolute_import
+
+
+import tornado
+import tornado.gen
+
+from ..messages.common import StreamState, FlagsType
+from .stream import InMemStream
+from .util import get_arg, serialize_json, deserialize_json
+from ..zipkin.trace import Trace
+from ..exceptions import TChannelException
+
+
+class Request(object):
+    """Represents an incoming request to an endpoint.
+
+    Request class is used to represent the CallRequestMessage at User's level.
+    This is going to hide the protocol level message information.
+    """
+
+    # TODO decide which elements inside "message" object to expose to user.
+    def __init__(
+        self,
+        id=None,
+        flags=FlagsType.none,
+        ttl=10,
+        tracing=None,
+        service=None,
+        headers=None,
+        checksum=None,
+        argstreams=None
+    ):
+        self.flags = flags
+        self.ttl = ttl
+        self.service = service
+        self.tracing = tracing or Trace()
+        # argstreams is a list of InMemStream/PipeStream objects
+        self.argstreams = argstreams
+        self.checksum = checksum
+        self.id = id
+        self.headers = headers or {}
+        self.state = StreamState.init
+        self.endpoint = ""
+        self.header = None
+        self.body = None
+
+    @property
+    def arg_scheme(self):
+        return self.headers.get('as', None)
+
+    def close_argstreams(self, force=False):
+        for stream in self.argstreams:
+            if stream.auto_close or force:
+                stream.close()
+
+    def get_header(self):
+        """Get the header value from the request.
+
+        :return: a future contains the value of header
+        """
+        return get_arg(self, 1)
+
+    def get_body(self):
+        """Get the body value from the request.
+
+        :return: a future contains the value of body
+        """
+        return get_arg(self, 2)
+
+    def get_header_s(self):
+        """Get the raw stream of header.
+
+        :return: the argstream of header
+        """
+        return self.argstreams[1]
+
+    def get_body_s(self):
+        """Get the raw stream of body.
+
+        :return: the argstream of body
+        """
+        return self.argstreams[2]
+
+
+class JsonRequest(Request):
+    def __init__(self, req=None, serializer=None, deserializer=None):
+        if isinstance(req, Request):
+            super(JsonRequest, self).__init__(
```

Comment:
Or perhaps a `RequestBuilder.from_json()` - in other-words, there is no benefit to this inheritance tree, it just makes the program brittle.